sklearn/feature_extraction/text.py

```diff
@@ -1078,7 +1078,8 @@ def fit(self, X, y=None):
 
             # log+1 instead of log makes sure terms with zero idf don't get
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
+            # astype will fallback to np.float64 if X.dtype is int
+            idf = np.log(float(n_samples) / df).astype(X.dtype) + 1.0
```

Comment:
We don't preserve backward compatibility if X is int (it is true that we keep the type thought).

```python
In [6]: np.log(5. / np.array([1, 2, 3.5])).astype(np.int32) + 1.0
Out[6]: array([ 2.,  1.,  1.])

In [7]: np.log(5. / np.array([1, 2, 3.5])) + 1.0
Out[7]: array([ 2.60943791,  1.91629073,  1.35667494])
```

The results are really different due to the casting.

Reply:
By the way I am not sure that `float(n_samples)` is useful `n_samples` should be enough.

@rth Could you confirm that my thoughts are right (I don't know the code base of the TD-IDF) in fact.